### PR TITLE
discv4: use ExpectedResponseChannels for neighbours msgs

### PR DESCRIPTION
**Pull request by gsalgado**
Part of #674

### CODE REVIEW

File: p2p/discovery.py [hunk 1]
```diff
@@ -337,33 +340,28 @@ def update_routing_table(self, node: NodeAPI) -> None:
...
+                # mypy thinks wrapping our generator turns it into something else, so ignore.
+                async with aclosing(gen):  # type: ignore
+                    async for batch in gen:
+                        self.logger.debug2(
+                            f'got expected neighbours response from {remote}: {batch}')
+                        neighbours.extend(batch)
+                        if len(neighbours) >= constants.KADEMLIA_BUCKET_SIZE:
+                            break
+                self.logger.debug2(f'got expected neighbours response from {remote}')
+        except trio.TooSlowError:
```
Comment (pipermerriam) [inline p2p/discovery.py]: I think the following format might be more readable.

```python
with trio.move_on_after(...) as cancel_scope:
    ...
if cancel_scope.cancelled_caught:
    self.logger.debug(...)
```

Just a slight reduction in levels of indention.  It does have the slight downside of allowing for the silent failure if the `if cancel_scope.cancelled_caught` block is accidentally removed, but I think I prefer this over the large `try` block.

File: p2p/discovery.py [hunk 2]
```diff
@@ -1110,20 +1102,53 @@ def already_waiting_for(self, remote: NodeAPI) -> bool:
...
+        on the given receive channel.
+
+        Closes the channel once a message is received.
+
+        Returns the read message.
+        """
+        gen = self.receive(remote, send_chan, recv_chan)
+        # mypy thinks wrapping our generator turns it into something else, so ignore.
+        async with aclosing(gen):  # type: ignore
+            return await gen.__anext__()
```
Comment (pipermerriam) [inline p2p/discovery.py]: Thoughts on doing this as:

```python
async for msg in self.receive(...):
    return msg
```

This should have the same effect of returning after the first value...

Comment (gsalgado) [reply p2p/discovery.py]: That's how I initially wrote it, but then mypy complains that the method is missing a return statement. Is there a way around that?

File: p2p/discovery.py (hunk 1 above)
Comment (gsalgado) [reply p2p/discovery.py]: Yeah, I think I also prefer this over the large try and extra level of indentation

File: p2p/discovery.py (hunk 2 above)
Comment (pipermerriam) [reply p2p/discovery.py]: I think this works:

```python
async for msg in self.receive(...):
    break
return msg
```

Comment (pipermerriam) [reply p2p/discovery.py]: it might also work to do `return await self.receive(...).asend(None)` but I'm not 100% sure that statement is equivalent....

Comment (gsalgado) [reply p2p/discovery.py]:
```
        gen = self.receive(remote, send_chan, recv_chan)
        async with aclosing(gen):  # type: ignore
            async for msg in gen:
                return msg
```

> p2p/discovery.py:1104: error: Missing return statement

**Pull request merged**